UART RTT devicetree & async support

JordanYates opened the pull request:
Switches instantiation of the `uart_rtt` driver from Kconfig symbols to devicetree nodes.
Adds `CONFIG_UART_ASYNC_API` support to the driver, which eliminates hard faults when the RTT backend is used with async functions. Because asynchronous reception is not possible, only the TX path is implemented.

There are no in tree users of the previous `CONFIG_UART_RTT_*` symbols, so there are no tests validating the new behavior. The commit message contains an example of the devicetree nodes required to instantiate driver instances.

Commit:
drivers: serial: uart_rtt converted to devicetree

Instantiate RTT UART instances from devicetree nodes instead of from
Kconfig symbols. While RTT is implemented using software, not hardware,
it is emulating a hardware device, and thus should be configured through
devicetree. This allows the simulated UART device to be selected via
devicetree aliases and chosen nodes.

The following devicetree snippet will instantiate RTT channels 0 and 2
as UART devices.
```
/ {
	rtt0: rtt_terminal {
		compatible = "segger,rtt-uart";
		label = "rtt_terminal";
		status = "okay";
	};

	rtt2: rtt_secondary {
		compatible = "segger,rtt-uart";
		label = "rtt_app_specific";
		status = "okay";
	};
};
```

Fixes the RTT portion of #10621.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
drivers: serial: uart_rtt async support

Add support for `CONFIG_UART_ASYNC_API` to the uart_rtt driver.
As RTT provides no mechanism for knowing when new data has arrived,
the reception commands all return errors.

This fixes hard faults when asynchronous calls are made on the uart_rtt
driver, in addition to providing a small level of functionality.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

henrikbrixandersen commented:
What is the real benefit of this? RTT is not a hardware device, it is purely software.

JordanYates commented:
> What is the real benefit of this? RTT is not a hardware device, it is purely software.

The benefit is that it allows the use of RTT in chosen and alias nodes.
I agree that the implementation is pure software, but it is pretending to be a hardware device, and therefore is seems reasonable (to me at least) to specify in devicetree.

pdunaj commented on drivers/serial/uart_rtt.c:
```diff
@@ -72,49 +86,140 @@ static void uart_rtt_poll_out(struct device *dev, unsigned char c)
+	struct uart_rtt_data *data = get_dev_data(dev);
+	unsigned int ch = cfg ? cfg->channel : 0;
+
+	ARG_UNUSED(timeout);
```
If timeout is not supported maybe tx should check for timeout being K_FOREVER and return an error in any other case?

JordanYates replied:
The docs for `uart_tx` say that the timeout parameter is only valid when flow control is enabled.
Given that flow control is never enabled, there doesn't seem to be a reason to do anything with a permanently invalid parameter.

pdunaj commented on drivers/serial/uart_rtt.c:
```diff
@@ -72,49 +86,140 @@ static void uart_rtt_poll_out(struct device *dev, unsigned char c)
+static int uart_rtt_rx_enable(struct device *dev, uint8_t *buf, size_t len,
+			      int32_t timeout)
+{
+	/**
```
convert to normal comment s/**/*

pdunaj commented on drivers/serial/uart_rtt.c:
```diff
@@ -72,49 +86,139 @@ static void uart_rtt_poll_out(struct device *dev, unsigned char c)
+	ARG_UNUSED(timeout);
+
+	/* Output the buffer */
+	SEGGER_RTT_Write(ch, buf, len);
```
`SEGGER_RTT_Write` is taking a lock (mutex in case of zephyr). This can limit places where this driver can be used.
Does Zephyr allow calling uart_tx from atomic context?

JordanYates replied:
I suspect that its not defined, and probably won't be until #18970 is resolved.
The nrfx and sam0 drivers are the only ones that currently implement `UART_ASYNC_API`.
They both look like they will work/return `-EBUSY` with interrupts disabled.
I will update to manually take the lock.

JordanYates replied:
Updated

pdunaj replied:
Not a bad idea but it will cover only cases when uart_tx is called from under irq_lock. It won't cover the case when it is called from an interrupt (mutex cannot be taken from an interrupt).

Maybe it would be enough to leave `SEGGER_RTT_Write` and add a check blocking usage from atomic context?

there is no is_atomic api in Zephyr but check can be done as
```
        int key = irq_lock();                                                                                            
        irq_unlock(key);                                                                                                 
                                                                                                                         
        bool is_atomic = arch_irq_unlocked(key) || k_is_in_isr();
        if (is_atomic) return -ENOTSUP;
```

JordanYates replied:
I'd rather not fail when IRQ's are locked if the mutex take would still succeed.
Is adding the following sufficient?
```
if (k_in_in_isr()) {
    return -ENOTSUP;
}
```

pdunaj replied:
That's ok. But I strongly suggest adding a check against calling from an interrupt. If anybody does it at least problem will be visible and explicit.

JordanYates replied:
Doesn't this setup cover the cases?
If in ISR the `k_in_in_isr` will fail.
If interrupts are locked in a thread the TX will pass if mutex is free, fail if it is in use.
Otherwise it will pass.
Is there some other situation I'm missing?

pdunaj replied:
No, it's fine :)

henrikbrixandersen commented:
I would still like a discussion on if this really is the right solution here.

I often find myself enabling the RTT console as a pure software configuration change while debugging. It would be quite bothersome to have to edit the devicetree for this to work. Is there any way we could support both methods? After all, this _is_ pure software functionality.

JordanYates commented:
> I often find myself enabling the RTT console as a pure software configuration change while debugging. It would be quite bothersome to have to edit the devicetree for this to work. Is there any way we could support both methods? After all, this _is_ pure software functionality.

My understanding is that changes to the uart_rtt driver are completely independent of the RTT console. `log_backend_rtt.c` calls the SEGGER functions directly, as does the char_out function provided to printk in `rtt_console.c`. This PR therefore shouldn't change anything about enabling RTT for debug.

JordanYates commented:
@galak @carlescufi any chance of another review, hoping to get this into 2.4

nordic-krch commented on drivers/serial/uart_rtt.c:
```diff
@@ -72,49 +86,152 @@ static void uart_rtt_poll_out(struct device *dev, unsigned char c)
+	}
+
+	/* Claim the RTT lock */
+	if (k_mutex_lock(&rtt_term_mutex, K_NO_WAIT) != 0) {
```
should you do `k_mutex_lock(&rtt_term_mutex, k_is_in_isr() ? K_NO_WAIT : K_FOREVER)`

JordanYates replied:
I would have thought so as well, but it is not valid to try locking a mutex inside an ISR at all, even with K_NO_WAIT.
https://github.com/zephyrproject-rtos/zephyr/blob/master/kernel/mutex.c#L125

nordic-krch reviewed (commented):
I think there is a zephyr policy that code must be somehow validated in the tree (compiled). Is it used somewhere?

JordanYates commented:
> I think there is a zephyr policy that code must be somehow validated in the tree (compiled). Is it used somewhere?

uart_rtt was previously not used anywhere in tree. I can whip up a sample application.

Commit:
tests: uart_async_api: build segger rtt driver

Add a build only test to validate correct compilation of the uart_rtt
driver. No harness that connects to RTT is defined, and therefore the
test can't be run by CI.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

JordanYates commented:
> I think there is a zephyr policy that code must be somehow validated in the tree (compiled). Is it used somewhere?

Pretty sure the testcase I added will compile for the 'qemu_cortex_m0' platform at least. (It does locally for me)

nordic-krch commented:
don't you need overlay which will configure uart rtt?

JordanYates commented:
> don't you need overlay which will configure uart rtt?

Thats the segger_rtt.overlay

nordic-krch commented:
> Thats the segger_rtt.overlay

somehow i missed that.
